rust/kernel/print.rs

```diff
@@ -0,0 +1,775 @@
+        assert!(bindings::KERN_CONT[2] == b'\x00');
+
+        const SUFFIX: &[u8; LENGTH_CONT - LENGTH_PREFIX] = b"%.*s\0";
+        [
+            bindings::KERN_CONT[0],
+            bindings::KERN_CONT[1],
+            SUFFIX[0],
+            SUFFIX[1],
+            SUFFIX[2],
+            SUFFIX[3],
+            SUFFIX[4],
+        ]
+    }
+
+    // Generate the format string at compile-time for the `CONT` "level".
+    //
+    // This avoids the compiler generating the contents on the fly in the stack.
+    //
+    // Furthermore, `static` instead of `const` is used to share the string
+    // for all the kernel.
+    pub static CONT: [u8; LENGTH_CONT] = generate_cont();
+}
+
+/// Prints a message via the kernel's [`printk`].
+///
+/// Hidden because it is called from a public macro -- it should never be
+/// called by users.
+///
+/// # Safety
+///
+/// The format string must be one of the ones in [`format_strings`].
+///
+/// (This ensures that the format string matches the variadic arguments we pass
+/// to [`printk`]).
+///
+/// [`printk`]: ../../../../include/linux/printk.h
+#[doc(hidden)]
+pub unsafe fn call_printk(
+    format_string: &[u8; format_strings::LENGTH],
+    module_name: &[u8],
+    string: &[u8],
+) {
+    // `printk` does not seem to fail in any path.
+    bindings::printk(
+        format_string.as_ptr() as _,
+        module_name.len() as c_int,
+        module_name.as_ptr(),
+        string.len() as c_int,
+        string.as_ptr(),
+    );
+}
+
+/// Prints a message via the kernel's [`printk`] for the `CONT` "level".
+///
+/// Follows the interface and qualifiers of [`call_printk`] to make things
+/// easier in [`print_macro!`].
+///
+/// Hidden because it is called from a public macro -- it should never be
+/// called by users.
+///
+/// # Safety
+///
+/// See [`call_printk`]'s safety contract.
+///
+/// [`printk`]: ../../../../include/linux/printk.h
+#[doc(hidden)]
+pub unsafe fn call_printk_cont(
+    format_string: &[u8; format_strings::LENGTH_CONT],
+    _module_name: &[u8],
+    string: &[u8],
+) {
+    // `printk` does not seem to fail in any path.
+    bindings::printk(
+        format_string.as_ptr() as _,
+        string.len() as c_int,
+        string.as_ptr(),
+    );
+}
+
+// From `kernel/print/printk.c`.
+const LOG_LINE_MAX: usize = 1024 - 32;
+
+#[doc(hidden)]
+pub struct LogLineWriter {
+    data: [u8; LOG_LINE_MAX],
+    pos: usize,
+}
+
+impl LogLineWriter {
+    /// Creates a new [`LogLineWriter`].
+    pub fn new() -> LogLineWriter {
+        LogLineWriter {
+            data: [0u8; LOG_LINE_MAX],
+            pos: 0,
+        }
+    }
+
+    /// Returns the internal buffer as a byte slice.
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.data[..self.pos]
+    }
+}
+
+impl Default for LogLineWriter {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl fmt::Write for LogLineWriter {
+    fn write_str(&mut self, s: &str) -> fmt::Result {
```

Comment:
Should change [`Buffer`](https://github.com/Rust-for-Linux/linux/blob/94450572762f32f55390609b3b4861d3bff0e244/rust/kernel/buffer.rs#L30) so that it can also be used here?

Reply:
Or perhaps creating a local array here, and then using `Buffer` on top of it.

But perhaps in another PR, since this one keeps what we were doing before.